### PR TITLE
feat(network): add Request.sizes()

### LINKED ISSUE #4469: [Feature] Report page assets size
Hi, 

It would be great if we can get transferred assets size with playwright api also the sizes can be grouped by request type.

When i use page.on('response') event, i can't get real encoded body size with this.

similiar repo: https://github.com/andersonba/puppeteer-assets

Thanks.

### PR DESCRIPTION
**Pull request by mxschmitt**
Fixes #4469

### CODE REVIEW

File: src/client/network.ts [hunk 1]
```diff
@@ -239,6 +251,18 @@ export class InterceptedResponse implements api.Response {
   request(): Request {
     return this._request;
   }
+
+  async headersSize(): Promise<number> {
```
Comment (mxschmitt) [inline src/client/network.ts:255]: not sure about this class, need to look into it.

File: src/server/webkit/wkPage.ts [hunk 2]
```diff
@@ -1051,7 +1051,7 @@ export class WKPage implements PageDelegate {
         validTo: responseReceivedPayload?.response.security?.certificate?.validUntil,
       });
       const { responseBodyBytesReceived, responseHeaderBytesReceived } = event.metrics || {};
-      const transferSize = responseBodyBytesReceived ? responseBodyBytesReceived + (responseHeaderBytesReceived || 0) : undefined;
+      const transferSize = responseBodyBytesReceived !== undefined ? responseBodyBytesReceived + (responseHeaderBytesReceived || 0) : undefined;
```
Comment (mxschmitt) [inline src/server/webkit/wkPage.ts:1054]: fixes a bug when responseBodySize was 0.

File: src/client/network.ts [hunk 3]
```diff
@@ -165,6 +165,18 @@ export class Request extends ChannelOwner<channels.RequestChannel, channels.Requ
     return this._timing;
   }
 
+  async bodySize(): Promise<number> {
```
Comment (pavelfeldman) [inline src/client/network.ts:168]: Should we expose request.sizes instead next to timing to keep it easily accessible?

File: src/server/network.ts [hunk 4]
```diff
@@ -193,6 +193,22 @@ export class Request extends SdkObject {
       this._headersMap.set('host', host);
     }
   }
+
+  bodySize(): number {
+    const postData = this.postDataBuffer();
+    if (!postData)
+      return 0;
+    return new TextEncoder().encode(postData.toString('utf8')).length;
```
Comment (pavelfeldman) [inline src/server/network.ts:201]: Why do we encode utf8 here? I would imagine we just take the buffer size.

File: src/server/network.ts [hunk 5]
```diff
@@ -193,6 +193,22 @@ export class Request extends SdkObject {
...
+  }
+
+  headersSize(): number {
+    if (!this._response)
+      return 0;
+    let rawHeaders = `${this._method} ${(new URL(this.url())).pathname} ${this._response.httpVersion()}\r\n`;
+    for (const header of this._headers)
+      rawHeaders += `${header.name}: ${header.value}\r\n`;
```
Comment (pavelfeldman) [inline src/server/network.ts:209]: You don't need to allocate lots of memory to sum up the size.

File: src/server/network.ts [hunk 6]
```diff
@@ -401,6 +417,32 @@ export class Response extends SdkObject {
...
+  httpVersion(): string {
+    if (!this._httpVersion)
+      return 'HTTP/1.1';
+    if (this._httpVersion === 'http/1.1')
+      return 'HTTP/1.1';
+    return this._httpVersion;
+  }
+
+  headersSize(): number {
+    let rawHeaders = `${this.httpVersion()} ${this.status()} ${this.statusText()}\r\n`;
```
Comment (pavelfeldman) [inline src/server/network.ts:440]: ditto.

Unfortunately this is all approximation, I wonder if users are going to be happy about it. It'll be not worse than in DevTools, but different from real network traffic due to headers encoding.

File: docs/src/api/class-request.md [hunk 7]
```diff
@@ -16,6 +16,11 @@ with `'requestfinished'` event.
 If request gets a 'redirect' response, the request is successfully finished with the 'requestfinished' event, and a new
 request is  issued to a redirected url.
 
+## method: Request.bodySize
```
Comment (pavelfeldman) [inline docs/src/api/class-request.md:19]: I remember suggesting Request.sizes next to Request.timing

File: src/protocol/protocol.yml [hunk 8]
```diff
@@ -731,6 +731,9 @@ BrowserContext:
       parameters:
         request: Request
         responseEndTiming: number
+        responseTransferSize: number
```
Comment (pavelfeldman) [inline src/protocol/protocol.yml:734]: We should use intermediate structs, can be RequestSizes for example and we can update / augment them incrementally, but should be a separate struct.

File: docs/src/api/class-request.md [hunk 9]
```diff
@@ -183,6 +183,17 @@ following: `document`, `stylesheet`, `image`, `media`, `font`, `script`, `texttr
 
 Returns the matching [Response] object, or `null` if the response was not received due to error.
 
+## method: Request.sizes
+- returns: <[Object]>
+  - `requestBodySize` <[int]> Size of the request body (POST data payload) in bytes. Set to 0 if there was no body.
+  - `requestHeadersSize` <[float]> Total number of bytes from the start of the HTTP request message until (and including) the double CRLF before the body.
+  - `responseBodySize` <[int]> Size of the received response body in bytes.
```
Comment (pavelfeldman) [inline docs/src/api/class-request.md:190]: Do we have a good idea on why users could want original body size? And why they don't want the actual transfer size for the request body?

File: src/server/network.ts [hunk 10]
```diff
@@ -401,6 +430,35 @@ export class Response extends SdkObject {
...
   }
+
+  transferSize(): number | undefined {
+    return this._transferSize;
+  }
+
+  bodySize(): number | undefined {
+    if (!this._transferSize)
+      return;
+    return this._transferSize - this.headersSize();
```
Comment (pavelfeldman) [inline src/server/network.ts:441]: You are going to see some negative sizes here.

File: src/server/network.ts [hunk 11]
```diff
@@ -193,6 +193,35 @@ export class Request extends SdkObject {
...
+    return postData.length;
+  }
+
+  headersSize(): number {
+    if (!this._response)
+      return 0;
+    let headersSize = 4; // 4 = 2 spaces + 2 line breaks (GET /path \r\n)
+    headersSize += this.method().length;
+    headersSize += (new URL(this.url())).pathname.length;
+    headersSize += this._response.httpVersion().length;
```
Comment (pavelfeldman) [inline src/server/network.ts:210]: Just add 7

File: src/server/network.ts [hunk 12]
```diff
@@ -401,6 +430,35 @@ export class Response extends SdkObject {
...
+      return 'HTTP/1.1';
+    if (this._httpVersion === 'http/1.1')
+      return 'HTTP/1.1';
+    return this._httpVersion;
+  }
+
+  headersSize(): number {
+    let headersSize = 4; // 4 = 2 spaces + 2 line breaks (HTTP/1.1 200 Ok\r\n)
+    headersSize += this.httpVersion().length;
+    headersSize += this.status().toString().length;
```
Comment (pavelfeldman) [inline src/server/network.ts:455]: I'm going to predict this is 3

File: tests/config/checkCoverage.js [hunk 13]
```diff
@@ -51,6 +51,11 @@ if (browserName !== 'chromium') {
 if (browserName === 'webkit')
   api.delete('browserContext.clearPermissions');
 
+// WebKit on Linux has no sizes caused by libsoup, waiting for libsoupv3
+if (browserName === 'webkit') {
+  api.delete('request.sizes');
```
Comment (pavelfeldman) [inline tests/config/checkCoverage.js:56]: I think we have body / transfer sizes through dataReceived in WebKit, don't we?

Comment (mxschmitt) [reply tests/config/checkCoverage.js:56]: Not on Linux unfortunately, there it's blocked by libsoup3.

File: src/server/network.ts (hunk 10 above)
Comment (mxschmitt) [reply src/server/network.ts:441]: In theory transfer size is always bigger, what would you change here?

File: docs/src/api/class-request.md (hunk 9 above)
Comment (mxschmitt) [reply docs/src/api/class-request.md:190]: It's from the HAR spec and I thought it's about exposing these values.

Comment (pavelfeldman) [reply docs/src/api/class-request.md:190]: You should look at har spec and re-read my question!

File: src/server/network.ts (hunk 10 above)
Comment (pavelfeldman) [reply src/server/network.ts:441]: The logic that you use when computing sizes is fuzzy, so you should tolerate inaccurate numbers, capping this to zero. Does transfer size include headers?

File: tests/config/checkCoverage.js (hunk 13 above)
Comment (pavelfeldman) [reply tests/config/checkCoverage.js:56]: So no data size in dataReceived?

**Review by pavelfeldman — APPROVED** [review]

**Pull request merged**